# Assignment 2, Question 2
by Zach Justus

## A. Import Libraries and the Dataset

In [1]:
# Import modules
import numpy as np
import pandas as pd

from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# Import the data
iris_df = pd.read_csv('https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv')
feature_cols = ['sepal_length','sepal_width','petal_length','petal_width']
X = iris_df[feature_cols]
y = iris_df['species']

## B. Split the data set into testing and training
test_size = 0.4
random_state = 6

In [2]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)

# C. Create KNN Object, Train, Then Test

In [3]:
# Create Object
k = 3
my_knn_for_cs4661 = KNeighborsClassifier(n_neighbors=k)

# Train Object
my_knn_for_cs4661.fit(X_train, y_train)

# Test Object
y_predict = my_knn_for_cs4661.predict(X_test)
accuracy = accuracy_score(y_test, y_predict)

print(accuracy)

0.9666666666666667


## D. Redo C with various k values
k = [1, 5, 7, 11, 15, 27, 59]

Q. Does the accuracy always get better by increasing the value K?  
A. No, the results of the code below show that there exists a particular value of k that yields the best accuracy for the data. going further away from that value the accuracy will decrease.

In [4]:
kTests = [1, 5, 7, 11, 15, 27, 59]

for k in kTests:
    my_knn_for_cs4661 = KNeighborsClassifier(n_neighbors=k)
    my_knn_for_cs4661.fit(X_train, y_train)
    y_predict = my_knn_for_cs4661.predict(X_test)
    accuracy = accuracy_score(y_test, y_predict)
    print("K = {}: {}".format(k, accuracy))

K = 1: 0.95
K = 5: 0.9833333333333333
K = 7: 0.9666666666666667
K = 11: 0.9666666666666667
K = 15: 0.9333333333333333
K = 27: 0.9166666666666666
K = 59: 0.8166666666666667


## E. Redo C but only make predictions using a single feature, do this for all features
Q. Which feature is the single best feature to use for prediction?  
A. petal_length is the best feature with a 96.7% accuracy

In [5]:
k = 3
my_knn_for_cs4661 = KNeighborsClassifier(n_neighbors=k)
for feature in X_train.columns:
    my_knn_for_cs4661.fit(X_train[[feature]], y_train)
    
    y_predict = my_knn_for_cs4661.predict(X_test[[feature]])
    accuracy = accuracy_score(y_test, y_predict)
    print("Feature = {}: {}".format(feature, accuracy))

Feature = sepal_length: 0.7166666666666667
Feature = sepal_width: 0.43333333333333335
Feature = petal_length: 0.9666666666666667
Feature = petal_width: 0.95


## F. Redo E but with predictions based on two data sets.
Q. Which feature pair is the best to use for prediction?  
A. There is a tie between the pair special_length & petal_length, and special_width & petal_length. They both have an accuracy of 98.3%

In [6]:
k = 3
my_knn_for_cs4661 = KNeighborsClassifier(n_neighbors=k)
fPairs = [[0,1], [0,2], [0,3], [1,2], [1,3], [2,3]]

for fp in fPairs:
    my_knn_for_cs4661.fit(X_train[[X_train.columns[fp[0]], X_train.columns[fp[1]]]], y_train)
    y_predict = my_knn_for_cs4661.predict(X_test[[X_train.columns[fp[0]], X_train.columns[fp[1]]]])
    accuracy = accuracy_score(y_test, y_predict)
    print("Features {} & {}: {}".format(X_train.columns[fp[0]], X_train.columns[fp[1]], accuracy))

Features sepal_length & sepal_width: 0.7833333333333333
Features sepal_length & petal_length: 0.9833333333333333
Features sepal_length & petal_width: 0.95
Features sepal_width & petal_length: 0.9833333333333333
Features sepal_width & petal_width: 0.95
Features petal_length & petal_width: 0.9666666666666667


# G. Big Picture
Q. Does the best feature pair from part F match the two best single features in part E?  
A. Yes and no, The best best feature in part E, Petal Length, Also happens to be a common feature in both ties in part F. however, the second best feature in part E, petal_width, has no relation to either of the two best pairs from section F. 

# H. Explaining G? 
I believe the reason why the 2nd most prominent feature of part E is not found at the top of part F because this feature has a data set too similar to the most prominent feature, the reason why the 3rd and 4th feature produces a better pair in part F is because when combined with the most predominant feature the set provides deeper diversity and insight into the correct outcomes of the data